Network/Socket/Types.hsc

```diff
@@ -70,14 +73,47 @@ import Network.Socket.Imports
 -----------------------------------------------------------------------------
 
 -- | Basic type for a socket.
-newtype Socket = Socket CInt deriving (Eq, Show)
+newtype Socket = Socket (IORef CInt)
+
+instance Show Socket where
+    show s = "<socket " ++ show (unsafeFdSocket s) ++ ">"
+
+instance Eq Socket where
+    Socket ref1 == Socket ref2 = ref1 == ref2
 
 -- | Getting a file descriptor from a socket.
-fdSocket :: Socket -> CInt
-fdSocket (Socket fd) = fd
+fdSocket :: Socket -> IO CInt
+fdSocket (Socket ref) = readIORef ref
+
+unsafeFdSocket :: Socket -> CInt
+unsafeFdSocket = unsafePerformIO . fdSocket
+{-# NOINLINE unsafeFdSocket #-}
+
+-- | Creating a socket form a file descriptor.
+mkSocket :: CInt -> IO Socket
+mkSocket fd = do
+    ref <- newIORef fd
+    let s = Socket ref
+    void $ mkWeakIORef ref $ close s
+    return s
+
+-----------------------------------------------------------------------------
+
+-- | Close the socket. Sending data to or receiving data from closed socket
+--   may lead to undefined behaviour.
+close :: Socket -> IO ()
+close (Socket ref) = do
+    oldfd <- atomicModifyIORef' ref $ \cur -> (-1, cur)
+    when (oldfd /= -1) $
+        closeFdWith (void . c_close . fromIntegral) (fromIntegral oldfd)
```

Comment:
Do we want to throw an error when attempting to close an already closed socket? I can see the argument for both ends of this.
1. The user might want to know when they are closing a socket that has already been closed. This is available in the underlying api:
  > close() returns zero on success. On error, -1 is returned, and errno is set appropriately.
2. Running `close` on an already closed socket is a safe enough operation. Throwing an exception seems overly explosive.

Maybe we need `close'`, which returns a value indicating whether the socket has been closed or not. The possibilities are binary, but a `Bool` doesn't really carry enough context to be useful here.

Reply:
If this `close` is to be safe for finalization, it MUST NOT throw exceptions.  So if exceptions are to be thrown, there would need to be a second (need not be exported) `close'` for finalization.